mev-boost-rs/src/relay.rs

```diff
@@ -0,0 +1,144 @@
+use beacon_api_client::Client as BeaconClient;
+use ethereum_consensus::{crypto::Error as CryptoError, primitives::BlsPublicKey};
+use thiserror::Error;
+use url::Url;
+
+use mev_rs::blinded_block_provider::Client;
+
+#[derive(Clone, Debug)]
+pub struct RelayEndpoint {
+    url: Url,
+    public_key: BlsPublicKey,
+}
+
+#[derive(Debug, Error)]
+pub enum RelayUrlError {
+    #[error("{0}")]
+    Bls(#[from] CryptoError),
+    #[error("{0}")]
+    Hex(#[from] hex::FromHexError),
+    #[error("public key {0} missing '0x' hex prefix")]
+    Missing0xPrefix(String),
+    #[error("URL {0} missing public key username")]
+    MissingPublicKey(String),
+}
+
+impl TryFrom<Url> for RelayEndpoint {
+    type Error = RelayUrlError;
+
+    fn try_from(url: Url) -> Result<Self, Self::Error> {
+        let public_key = url.username();
+        if public_key.is_empty() {
+            return Err(Self::Error::MissingPublicKey(url.to_string()));
+        }
+
+        let public_key =
+            public_key.strip_prefix("0x").ok_or(Self::Error::Missing0xPrefix(public_key.into()))?;
+        let public_key = hex::decode(public_key)?;
+        let public_key = BlsPublicKey::try_from(public_key.as_slice())?;
```

Comment:
this would add a dep on `serde_json` but another way to relay on the serde logic already defined for `BlsPublicKey` (which is my preference)

something like

```suggestion
        let public_key_hex = url.username();
        let public_key: BlsPublicKey = serde_json::from_str(&public_key_hex)?;
```

the error handling may get a little weird converting to an acceptable type, but at the least we can use `ethereum_consensus::serde::try_bytes_from_hex_str` to keep all the logic in one place (and drop the explicit `hex` dep added here)